File: dropdeck/framing.py
"""Telling a presenter who cannot see the screen what the camera can see.

This is the part of camera streaming worth building. OBS answers "am I in
shot?" with a preview window, which is not an answer, it is the question
restated. A blind presenter going live has no way to know whether they are
centred, lit, or whether they wandered out of frame twenty minutes ago and have
been broadcasting an empty chair since.

So the app says it.

**Not repetitively, which is the whole design.** A face wobbles. A detector
that reports what it sees would say "centred, left, centred, left" for three
hours, over the top of a screen reader, on air. Three things stop that and all
three are needed:

1. **Changes are spoken, never states.** Nothing is said while the reading is
   what it already was. This removes almost all of it on its own.
2. **Hysteresis on every threshold.** A face sitting on a boundary has to cross
   a wider band to change the answer than it did to set it. Without this, one
   pixel of wobble flips the answer back and forth for ever.
3. **A floor between announcements.** Several seconds minimum, counted from
   when something was last SAID rather than from when it changed.

**On demand is the primary interface, and the announcements are the backstop.**
Somebody setting a shot up wants to ask over and over for ten seconds and then
never again, which is a key, not a running commentary. `describe()` always
answers, at every level, including when the detector is switched off.

The default level is PROBLEMS: quiet while the shot is good, speaks when you
leave it. A show is three hours and what you need to hear about is the twenty
seconds that went wrong.

The detector is YuNet, through OpenCV. It runs on a small greyscale copy two or
three times a second, not on every frame: measured 7 September 2026 on a real
camera, a face found in 36 of 36 checks at 3.0 ms on average and 12.6 ms at
worst. **OpenCV 5.0 has removed `CascadeClassifier`**, so there is no Haar
fallback; without the model the framing simply turns itself off and says so,
and streaming is unaffected.
"""
from __future__ import annotations

import os
import sys
import threading
import time

import numpy as np

from . import constants as C

try:
    import cv2
except Exception:      # pragma: no cover - OpenCV missing is a real state
    cv2 = None


#: How much of the framing is spoken. Not the app's speech levels: this is one
#: setting about one thing, and it defaults to the quiet useful middle.
FRAMING_OFF = "off"
FRAMING_PROBLEMS = "problems"
FRAMING_EVERYTHING = "everything"
FRAMING_LEVELS = (FRAMING_OFF, FRAMING_PROBLEMS, FRAMING_EVERYTHING)

FRAMING_LEVEL_LABELS = {
    FRAMING_OFF: "Do not tell me about the shot",
    FRAMING_PROBLEMS: "Tell me when something is wrong",
    FRAMING_EVERYTHING: "Tell me about every change",
}


def model_path():
    """Where the detector's model lives, frozen or from source."""
    base = getattr(sys, "_MEIPASS", None)
    if base is None:
        base = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    return os.path.join(base, "assets", "models", C.FACE_MODEL_FILE)


def available():
    """Whether framing can work at all here."""
    return cv2 is not None and hasattr(cv2, "FaceDetectorYN") \
        and os.path.isfile(model_path())


def why_unavailable():
    """The reason, for saying out loud. Empty when it is available."""
    if cv2 is None:
        return "this copy cannot look at the picture"
    if not hasattr(cv2, "FaceDetectorYN"):
        return "this copy has no face detector"
    if not os.path.isfile(model_path()):
        return "the face detection model is missing"
    return ""


# ---------------------------------------------------------------------------
# One look at the picture
# ---------------------------------------------------------------------------

class Reading:
    """What the camera can see, in words rather than numbers."""

    __slots__ = ("found", "horizontal", "vertical", "distance", "light",
                 "centre_x", "centre_y", "size", "luminance", "confidence")

    def __init__(self, found=False, horizontal="", vertical="", distance="",
                 light="", centre_x=0.0, centre_y=0.0, size=0.0,
                 luminance=0.0, confidence=0.0):
        self.found = found
        self.horizontal = horizontal
        self.vertical = vertical
        self.distance = distance
        self.light = light
        self.centre_x = centre_x
        self.centre_y = centre_y
        self.size = size
        self.luminance = luminance
        self.confidence = confidence

    @property
    def key(self):
        """What has to change before anything is worth saying again."""
        return (self.found, self.horizontal, self.vertical, self.distance,
                self.light)

    @property
    def good(self):
        """A shot nobody needs to be told about."""
        return (self.found and self.horizontal == "centred"
                and self.vertical == "centred" and self.distance != "far away"
                and self.light != "dark")

    def sentence(self):
        """The whole reading, for the key that answers on demand."""
        if not self.found:
            if self.light == "dark":
                return "No face in shot, and the picture is dark"
            return "No face in shot"
        parts = []
        if self.horizontal == "centred" and self.vertical == "centred":
            parts.append("centred")
        else:
            if self.horizontal != "centred":
                parts.append(self.horizontal)
            if self.vertical != "centred":
                parts.append(self.vertical)
        parts.append(self.distance)
        parts.append(self.light)
        return ", ".join(p for p in parts if p).capitalize()

    def problem(self):
        """The one thing worth interrupting for, or empty."""
        if not self.found:
            return "No face in shot"
        if self.light == "dark":
            return "The picture is dark"
        if self.horizontal != "centred":
            return self.horizontal.capitalize()
        if self.vertical != "centred":
            return self.vertical.capitalize()
        if self.distance == "far away":
            return "Far away from the camera"
        return ""


def _band(value, low, high, below, middle, above, previous="", margin=0.0):
    """One reading with hysteresis, so a wobble does not flip the answer.

    Coming OUT of a state needs `margin` more than going in did. Without it a
    face resting on a boundary changes the answer several times a second, and
    the announcement floor then hides real changes behind fake ones.
    """
    lo, hi = low, high
    if previous == below:
        lo = low + margin
    elif previous == above:
        hi = high - margin
    elif previous == middle:
        lo, hi = low - margin, high + margin
    if value < lo:
        return below
    if value > hi:
        return above
    return middle


class Framer:
    """Looks at frames and says what changed, without becoming a commentary."""

    def __init__(self, level=FRAMING_PROBLEMS, on_say=None, clock=None):
        self.level = level if level in FRAMING_LEVELS else FRAMING_PROBLEMS
        self.on_say = on_say or (lambda text: None)
        self._clock = clock or time.monotonic
        self._detector = None
        self._input = None
        self._lock = threading.Lock()
        self.error = ""
        self.checks = 0
        self.reading = Reading()
        self._said_key = None
        self._said_at = 0.0
        self._last_look = 0.0

    # ------------------------------------------------------------ detector --
    def _ensure(self):
        if self._detector is not None or self.error:
            return
        reason = why_unavailable()
        if reason:
            self.error = reason
            return
        try:
            self._detector = cv2.FaceDetectorYN.create(
                model_path(), "", (C.FACE_INPUT_WIDTH, C.FACE_INPUT_HEIGHT),
                score_threshold=C.FACE_CONFIDENCE)
            self._input = (C.FACE_INPUT_WIDTH, C.FACE_INPUT_HEIGHT)
        except Exception as exc:
            self.error = "the face detector would not start: %s" % exc

    # ---------------------------------------------------------- one look --
    def due(self):
        """Whether it is time to look again. Cheap, so it can be asked often."""
        return (self._clock() - self._last_look) >= C.FACE_CHECK_SECONDS

    def look(self, rgb):
        """Analyse one frame and speak if anything changed. Never raises."""
        self._last_look = self._clock()
        reading = self.measure(rgb)
        with self._lock:
            self.reading = reading
        self._maybe_say(reading)
        return reading

    def measure(self, rgb):
        """One reading, with no speech and no state. Safe to call anywhere."""
        self._ensure()
        if rgb is None:
            return Reading()
        try:
            luminance = float(rgb[::8, ::8].mean())
        except Exception:
            luminance = 0.0
        light = "dark" if luminance < C.FACE_DARK_BELOW else "well lit"
        if self._detector is None:
            return Reading(light=light, luminance=luminance)

        try:
            height, width = rgb.shape[:2]
            scale = C.FACE_INPUT_WIDTH / float(width or 1)
            small_h = max(2, int(round(height * scale)))
            small = cv2.resize(rgb, (C.FACE_INPUT_WIDTH, small_h))
            # The detector has to be told the shape it is being given, and a
            # 4:3 camera is not the shape a 16:9 one is.
            if self._input != (C.FACE_INPUT_WIDTH, small_h):
                self._detector.setInputSize((C.FACE_INPUT_WIDTH, small_h))
                self._input = (C.FACE_INPUT_WIDTH, small_h)
            bgr = cv2.cvtColor(small, cv2.COLOR_RGB2BGR)
            _count, faces = self._detector.detect(bgr)
            self.checks += 1
        except Exception:
            # A detector that throws must never take the show down. No reading
            # is a fine answer; a crashed stream is not.
            return Reading(light=light, luminance=luminance)

        if faces is None or not len(faces):
            return Reading(light=light, luminance=luminance)

        # The biggest face, which is the presenter. Anyone in the background
        # is smaller and is not who this is for.
        face = max(faces, key=lambda f: float(f[2]) * float(f[3]))
        x, y, w, h = (float(face[0]), float(face[1]),
                      float(face[2]), float(face[3]))
        confidence = float(face[-1])
        centre_x = (x + w / 2.0) / float(C.FACE_INPUT_WIDTH)
        centre_y = (y + h / 2.0) / float(small_h)
        size = w / float(C.FACE_INPUT_WIDTH)

        previous = self.reading
        horizontal = _band(
            centre_x, C.FACE_LEFT_EDGE, C.FACE_RIGHT_EDGE,
            "left of shot", "centred", "right of shot",
            previous.horizontal, C.FACE_HYSTERESIS)
        vertical = _band(
            centre_y, C.FACE_TOP_EDGE, C.FACE_BOTTOM_EDGE,
            "high in shot", "centred", "low in shot",
            previous.vertical, C.FACE_HYSTERESIS)
        distance = _band(
            size, C.FACE_FAR_BELOW, C.FACE_CLOSE_ABOVE,
            "far away", "a good distance", "very close",
            previous.distance, C.FACE_SIZE_HYSTERESIS)
        return Reading(True, horizontal, vertical, distance, light,
                       centre_x, centre_y, size, luminance, confidence)

    # -------------------------------------------------------------- speech --
    def _maybe_say(self, reading):
        if self.level == FRAMING_OFF:
            self._said_key = reading.key
            return
        if reading.key == self._said_key:
            return
        now = self._clock()
        first = self._said_key is None
        if not first and (now - self._said_at) < C.FACE_SAY_FLOOR:
            # Too soon. The key is deliberately NOT recorded, so the change is
            # still pending and gets said once the floor has passed rather
            # than being lost.
            return

        text = self._words(reading, first)
        self._said_key = reading.key
        if not text:
            return
        self._said_at = now
        try:
            self.on_say(text)
        except Exception:
            pass

    def _words(self, reading, first):
        """What to say about this reading, or nothing."""
        if self.level == FRAMING_EVERYTHING:
            return reading.sentence()
        # Problems only: speak a fault, and speak recovery ONCE, because
        # "you are back in shot" is the other half of "you have gone".
        problem = reading.problem()
        if problem:
            return problem
        if first:
            return ""
        return "Back in shot" if reading.found else ""

    # ------------------------------------------------------------ on demand --
    def describe(self):
        """Always answers, at every level, including off.

        This is the key a presenter actually uses, and a switch that silences
        the announcements must not silence the answer to a direct question.
        """
        if self.error:
            return self.error
        with self._lock:
            reading = self.reading
        if not self.checks:
            return "The camera has not been looked at yet"
        return reading.sentence()

    def reset(self):
        """Forget what was said, so a new shot starts clean."""
        with self._lock:
            self.reading = Reading()
        self._said_key = None
        self._said_at = 0.0

File: mac/Sources/Framing.swift
// What the camera can see, said out loud, without becoming a commentary.
//
// Mirrors dropdeck/framing.py: the same bands, the same hysteresis, the same
// sentences, the same anti-repetition rule. What differs is the detector.
// Windows loads a YuNet model through OpenCV and ships both; the Vision
// framework is already in the process here, so the Mac ships neither and gets
// a better detector.
//
// One consequence worth stating, because it removes a branch rather than
// adding one. On Windows the whole feature can be UNAVAILABLE, when the wheel
// failed to install or the model file is missing, and there is an error path
// through every function for it. Vision cannot be missing from a Mac, so
// framing here either has an answer or has not looked yet.
//
// **It says changes, not states.** At the most talkative setting, six seconds
// in front of a camera produced one sentence. A running commentary on your own
// face while you are trying to present a show is worse than silence, which is
// why there are three levels and why the floor between announcements exists.

import Foundation
import Vision
import CoreVideo

/// What the camera can see, in words rather than numbers.
struct FramingReading {
    var found = false
    var horizontal = ""
    var vertical = ""
    var distance = ""
    var light = ""
    var centreX = 0.0
    var centreY = 0.0
    var size = 0.0
    var luminance = 0.0
    var confidence = 0.0

    /// What has to change before anything is worth saying again.
    var key: String {
        "\(found)|\(horizontal)|\(vertical)|\(distance)|\(light)"
    }

    /// A shot nobody needs to be told about.
    var good: Bool {
        found && horizontal == "centred" && vertical == "centred"
            && distance != "far away" && light != "dark"
    }

    /// The whole reading, for the key that answers on demand.
    func sentence() -> String {
        if !found {
            return light == "dark" ? "No face in shot, and the picture is dark"
                                   : "No face in shot"
        }
        var parts: [String] = []
        if horizontal == "centred" && vertical == "centred" {
            parts.append("centred")
        } else {
            if horizontal != "centred" { parts.append(horizontal) }
            if vertical != "centred" { parts.append(vertical) }
        }
        parts.append(distance)
        parts.append(light)
        let joined = parts.filter { !$0.isEmpty }.joined(separator: ", ")
        // Python's str.capitalize lowercases the rest, and every band here is
        // already lower case, so the two agree.
        guard let first = joined.first else { return joined }
        return String(first).uppercased() + joined.dropFirst().lowercased()
    }

    /// The one thing worth interrupting for, or nothing.
    func problem() -> String {
        if !found { return "No face in shot" }
        if light == "dark" { return "The picture is dark" }
        if horizontal != "centred" { return capitalised(horizontal) }
        if vertical != "centred" { return capitalised(vertical) }
        if distance == "far away" { return "Far away from the camera" }
        return ""
    }

    private func capitalised(_ s: String) -> String {
        guard let first = s.first else { return s }
        return String(first).uppercased() + s.dropFirst().lowercased()
    }
}

/// Looks at frames and says what changed.
final class Framer {

    var level: String
    var onSay: (String) -> Void
    var clock: () -> Double

    private let lock = NSLock()
    private(set) var reading = FramingReading()
    private(set) var checks = 0
    private(set) var error = ""
    private var saidKey: String?
    private var saidAt: Double = 0
    private var lastLook: Double = 0

    init(level: String = C.framingProblems,
         onSay: @escaping (String) -> Void = { _ in },
         clock: @escaping () -> Double = { ProcessInfo.processInfo.systemUptime }) {
        self.level = C.framingLevels.contains(level) ? level : C.framingProblems
        self.onSay = onSay
        self.clock = clock
    }

    /// Whether it is time to look again. Cheap, so it can be asked often.
    func due() -> Bool { (clock() - lastLook) >= C.faceCheckSeconds }

    /// Analyse one frame and speak if anything changed.
    @discardableResult
    func look(_ buffer: CVPixelBuffer) -> FramingReading {
        lastLook = clock()
        let made = measure(buffer)
        lock.lock(); reading = made; lock.unlock()
        maybeSay(made)
        return made
    }

    /// One reading, with no speech and no state. Safe to call anywhere.
    func measure(_ buffer: CVPixelBuffer) -> FramingReading {
        let width = CVPixelBufferGetWidth(buffer)
        let height = CVPixelBufferGetHeight(buffer)
        guard width >= 2, height >= 2 else { return FramingReading() }

        let luminance = Framer.brightness(buffer)
        let light = luminance < C.faceDarkBelow ? "dark" : "well lit"

        let request = VNDetectFaceRectanglesRequest()
        let handler = VNImageRequestHandler(cvPixelBuffer: buffer, options: [:])
        do {
            try handler.perform([request])
        } catch {
            // A detector that fails must never take the show down. No reading
            // is a fine answer; a stopped stream is not.
            return FramingReading(light: light, luminance: luminance)
        }
        lock.lock(); checks += 1; lock.unlock()

        // The biggest face, which is the presenter. Anyone in the background
        // is smaller and is not who this is for.
        let faces = (request.results ?? []).filter { Double($0.confidence) >= C.faceConfidence }
        guard let face = faces.max(by: { $0.boundingBox.width * $0.boundingBox.height
                                       < $1.boundingBox.width * $1.boundingBox.height })
        else {
            return FramingReading(light: light, luminance: luminance)
        }

        // Vision's box is in a coordinate space with the origin at the BOTTOM
        // left and everything as a fraction of the image. The bands are
        // written the way a person describes a picture, from the top, so the
        // vertical is flipped here rather than in the bands.
        let box = face.boundingBox
        let centreX = Double(box.midX)
        let centreY = 1.0 - Double(box.midY)
        let size = Double(box.width)

        let previous = reading
        let horizontal = Framer.band(centreX, C.faceLeftEdge, C.faceRightEdge,
                                     "left of shot", "centred", "right of shot",
                                     previous.horizontal, C.faceHysteresis)
        let vertical = Framer.band(centreY, C.faceTopEdge, C.faceBottomEdge,
                                   "high in shot", "centred", "low in shot",
                                   previous.vertical, C.faceHysteresis)
        let distance = Framer.band(size, C.faceFarBelow, C.faceCloseAbove,
                                   "far away", "a good distance", "very close",
                                   previous.distance, C.faceSizeHysteresis)
        return FramingReading(found: true, horizontal: horizontal, vertical: vertical,
                              distance: distance, light: light, centreX: centreX,
                              centreY: centreY, size: size, luminance: luminance,
                              confidence: Double(face.confidence))
    }

    /// One reading with hysteresis, so a wobble does not flip the answer.
    ///
    /// Coming OUT of a state needs `margin` more than going in did. Without it
    /// a face resting on a boundary changes the answer several times a second,
    /// and the announcement floor then hides real changes behind fake ones.
    static func band(_ value: Double, _ low: Double, _ high: Double,
                     _ below: String, _ middle: String, _ above: String,
                     _ previous: String = "", _ margin: Double = 0) -> String {
        var lo = low, hi = high
        if previous == below { lo = low + margin }
        else if previous == above { hi = high - margin }
        else if previous == middle { lo = low - margin; hi = high + margin }
        if value < lo { return below }
        if value > hi { return above }
        return middle
    }

    /// The mean level of a frame, over the same subsample Windows takes.
    static func brightness(_ buffer: CVPixelBuffer) -> Double {
        CVPixelBufferLockBaseAddress(buffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(buffer, .readOnly) }
        guard let base = CVPixelBufferGetBaseAddress(buffer)?
                .assumingMemoryBound(to: UInt8.self) else { return 0 }
        let w = CVPixelBufferGetWidth(buffer), h = CVPixelBufferGetHeight(buffer)
        let stride = CVPixelBufferGetBytesPerRow(buffer)
        var total = 0, n = 0
        for y in Swift.stride(from: 0, to: h, by: 8) {
            for x in Swift.stride(from: 0, to: w, by: 8) {
                let p = base + y * stride + x * 4
                // BGRA, and the alpha is not part of how bright anything is.
                total += Int(p[0]) + Int(p[1]) + Int(p[2])
                n += 3
            }
        }
        return n > 0 ? Double(total) / Double(n) : 0
    }

    // ----------------------------------------------------------- speech ---

    /// Run the announcement rule over a reading, without a camera.
    ///
    /// Public because the checks drive it directly: the rule is the part a
    /// presenter actually experiences, and it is worth being able to test it
    /// against a scripted minute rather than against a face.
    func announce(_ reading: FramingReading) { maybeSay(reading) }

    /// Record a reading as the current one, without announcing it.
    func remember(_ reading: FramingReading) {
        lock.lock(); self.reading = reading; lock.unlock()
    }

    private func maybeSay(_ reading: FramingReading) {
        if level == C.framingOff { saidKey = reading.key; return }
        if reading.key == saidKey { return }
        let now = clock()
        let first = saidKey == nil
        if !first && (now - saidAt) < C.faceSayFloor {
            // Too soon. The key is deliberately NOT recorded, so the change is
            // still pending and gets said once the floor has passed rather
            // than being lost.
            return
        }
        let text = words(reading, first: first)
        saidKey = reading.key
        if text.isEmpty { return }
        saidAt = now
        onSay(text)
    }

    private func words(_ reading: FramingReading, first: Bool) -> String {
        if level == C.framingEverything { return reading.sentence() }
        // Problems only: speak a fault, and speak recovery ONCE, because "you
        // are back in shot" is the other half of "you have gone".
        let problem = reading.problem()
        if !problem.isEmpty { return problem }
        if first { return "" }
        return reading.found ? "Back in shot" : ""
    }

    /// Always answers, at every level, including off.
    ///
    /// This is the key a presenter actually uses, and a switch that silences
    /// the announcements must not silence the answer to a direct question.
    func describe() -> String {
        if !error.isEmpty { return error }
        lock.lock()
        let made = reading
        let count = checks
        lock.unlock()
        if count == 0 { return "The camera has not been looked at yet" }
        return made.sentence()
    }

    /// Forget what was said, so a new shot starts clean.
    func reset() {
        lock.lock(); reading = FramingReading(); lock.unlock()
        saidKey = nil
        saidAt = 0
    }
}
